### PR TITLE
deploy: support local bin path

### PR DESCRIPTION
**Pull request by wzru**

### CODE REVIEW

File: tidb/deploy.bash [hunk 1]
```diff
@@ -19,4 +19,56 @@ if [ "${skip_exist}" == 'true' ]; then
 	fi
 fi
 
+ver_path=`expr "${ver}" : '\(.*+\)' || true`
+if [ $ver_path ]; then
+    path=${ver#*+}
```
Comment (innerr) [inline tidb/deploy.bash:24]: quote with "", if it's const, quote with ''

File: tidb/deploy.bash [hunk 2]
```diff
@@ -19,4 +19,56 @@ if [ "${skip_exist}" == 'true' ]; then
...
+	path=
+fi
+
 tiup cluster deploy "${name}" "${ver}" "${yaml}"${confirm}
+
+if [[ "$OSTYPE" == "linux-gnu"* ]]; then
+    os=linux
+elif [[ "$OSTYPE" == "darwin"* ]]; then
+    os=darwin
+else
```
Comment (innerr) [inline tidb/deploy.bash:36]: not sure is win here, maybe just exit 1

File: tidb/deploy.bash [hunk 3]
```diff
@@ -19,4 +19,56 @@ if [ "${skip_exist}" == 'true' ]; then
...
+    path=${ver#*+}
+    ver=${ver_path%+}
+else
+	path=
+fi
+
 tiup cluster deploy "${name}" "${ver}" "${yaml}"${confirm}
+
+if [[ "$OSTYPE" == "linux-gnu"* ]]; then
+    os=linux
```
Comment (innerr) [inline tidb/deploy.bash:33]: quote with ''

File: tidb/deploy.bash [hunk 4]
```diff
@@ -19,4 +19,56 @@ if [ "${skip_exist}" == 'true' ]; then
...
+if [ $ver_path ]; then
+    path=${ver#*+}
+    ver=${ver_path%+}
+else
+	path=
+fi
+
 tiup cluster deploy "${name}" "${ver}" "${yaml}"${confirm}
+
+if [[ "$OSTYPE" == "linux-gnu"* ]]; then
```
Comment (innerr) [inline tidb/deploy.bash:32]: $OSTYPE => ${OSTYPE}

File: tidb/deploy.bash [hunk 5]
```diff
@@ -19,4 +19,56 @@ if [ "${skip_exist}" == 'true' ]; then
...
+    os=linux
+elif [[ "$OSTYPE" == "darwin"* ]]; then
+    os=darwin
+else
+    os=win
+fi
+
+case $(uname -m) in
```
Comment (innerr) [inline tidb/deploy.bash:40]: Put those into <repo>/helper/helper.bash

File: tidb/deploy.bash [hunk 6]
```diff
@@ -19,4 +19,56 @@ if [ "${skip_exist}" == 'true' ]; then
...
+    i386)   arch="386" ;;
+    i686)   arch="386" ;;
+    x86_64) arch="amd64" ;;
+    arm)    arch="arm64" ;;
+esac
+
+if [ ${path} ]; then
```
Comment (innerr) [inline tidb/deploy.bash:47]: quote with ""

File: tidb/deploy.bash [hunk 7]
```diff
@@ -19,4 +19,56 @@ if [ "${skip_exist}" == 'true' ]; then
...
+    tiup cluster start "${name}"
+    if [ -d "${path}" ]; then
+		cd ${path}
+        if [ -x "${path}/tidb-server" ]; then
+            tar -czvf tidb-nightly-${os}-${arch}.tar.gz tidb-server
```
Comment (innerr) [inline tidb/deploy.bash:52]: Not sure is nightly

File: tidb/deploy.bash [hunk 8]
```diff
@@ -19,4 +19,56 @@ if [ "${skip_exist}" == 'true' ]; then
...
+        base=`basename ${path}`
+		dir=`dirname ${path}`
+        role=${base%*-server}
+		cd ${dir}
+        tar -czvf tidb-nightly-${os}-${arch}.tar.gz ${base}
+        tiup cluster patch "${name}" ${role}-nightly-${os}-${arch}.tar.gz -R ${role} --yes
+		cd -
+    fi
+    tiup cluster stop "${name}" --yes
+fi
```
Comment (innerr) [inline tidb/deploy.bash:74]: /n here

File: tidb/deploy.bash [hunk 9]
```diff
@@ -19,4 +19,56 @@ if [ "${skip_exist}" == 'true' ]; then
...
+        if [ -x "${path}/tidb-server" ]; then
+            tar -czvf tidb-nightly-${os}-${arch}.tar.gz tidb-server
+            tiup cluster patch "${name}" tidb-nightly-${os}-${arch}.tar.gz -R tidb --yes
+        fi
+        if [ -x "${path}/tikv-server" ]; then
+            tar -czvf tikv-nightly-${os}-${arch}.tar.gz tikv-server
+            tiup cluster patch "${name}" tidb-nightly-${os}-${arch}.tar.gz -R tikv --yes
+        fi
+        if [ -x "${path}/pd-server" ]; then
+            tar -czvf pd-nightly-${os}-${arch}.tar.gz pd-server
```
Comment (innerr) [inline tidb/deploy.bash:60]: all used tools should be noted in `deploy.bash#[dep]`

File: tidb/deploy.bash [hunk 10]
```diff
@@ -19,4 +19,56 @@ if [ "${skip_exist}" == 'true' ]; then
...
+
+if [ ${path} ]; then
+    tiup cluster start "${name}"
+    if [ -d "${path}" ]; then
+		cd ${path}
+        if [ -x "${path}/tidb-server" ]; then
+            tar -czvf tidb-nightly-${os}-${arch}.tar.gz tidb-server
+            tiup cluster patch "${name}" tidb-nightly-${os}-${arch}.tar.gz -R tidb --yes
+        fi
+        if [ -x "${path}/tikv-server" ]; then
```
Comment (innerr) [inline tidb/deploy.bash:55]: -x => -f

File: tidb/deploy.bash [hunk 11]
```diff
@@ -19,4 +19,56 @@ if [ "${skip_exist}" == 'true' ]; then
 	fi
 fi
 
+ver_path=`expr "${ver}" : '\(.*+\)' || true`
```
Comment (innerr) [inline tidb/deploy.bash:22]: format: v5.1.0+path

File: tidb/deploy.bash [hunk 12]
```diff
@@ -19,4 +19,56 @@ if [ "${skip_exist}" == 'true' ]; then
...
+        fi
+        if [ -x "${path}/tikv-server" ]; then
+            tar -czvf tikv-nightly-${os}-${arch}.tar.gz tikv-server
+            tiup cluster patch "${name}" tidb-nightly-${os}-${arch}.tar.gz -R tikv --yes
+        fi
+        if [ -x "${path}/pd-server" ]; then
+            tar -czvf pd-nightly-${os}-${arch}.tar.gz pd-server
+            tiup cluster patch "${name}" tidb-nightly-${os}-${arch}.tar.gz -R pd --yes
+        fi
+		cd -
```
Comment (innerr) [inline tidb/deploy.bash:63]: old style way should be better:
old_dir=`pwd`
...
cd "${old_dir}"

File: tidb/deploy.bash (hunk 10 above)
Comment (wzru) [reply tidb/deploy.bash]: binary file should be executable

File: helper/helper.bash [hunk 13]
```diff
@@ -116,3 +116,28 @@ function config_mysql()
...
 }
+
+function must_get_os_arch()
+{
+	if [[ "${OSTYPE}" == 'linux-gnu'* ]]; then
+		os='linux'
+	elif [[ "${OSTYPE}" == 'darwin'* ]]; then
+		os='darwin'
+	else
+		echo "[:-] not support os ${OSTYPE}"
```
Comment (innerr) [inline helper/helper.bash:127]: echo "" >&2

File: helper/helper.bash [hunk 14]
```diff
@@ -116,3 +116,28 @@ function config_mysql()
...
+	    i686)   arch='386' ;;
+	    x86_64) arch='amd64' ;;
+	    arm)    arch='arm64' ;;
+	esac
+}
+
+function cluster_patch()
+{
+	local role="${1}"
+	tar -czvf "${role}-local-${os}-${arch}.tar.gz" ${role}-server
```
Comment (innerr) [inline helper/helper.bash:141]: "${role}-server"

File: helper/helper.bash [hunk 15]
```diff
@@ -116,3 +116,28 @@ function config_mysql()
...
+    tiup cluster patch "${name}" "${role}-local-${os}-${arch}.tar.gz" -R ${role} --yes --offline
```
Comment (innerr) [inline helper/helper.bash:142]: "${role}"

File: helper/helper.bash [hunk 16]
```diff
@@ -116,3 +116,28 @@ function config_mysql()
 	echo "    - mysql.port = ${port}"
 	echo "    - mysql.user = ${user}"
 }
+
+function must_get_os_arch()
+{
+	if [[ "${OSTYPE}" == 'linux-gnu'* ]]; then
+		os='linux'
```
Comment (innerr) [inline helper/helper.bash:123]: local os=''

File: tidb/deploy.bash.ticat [hunk 17]
```diff
@@ -22,3 +22,5 @@ tidb.op.confirm|cfm = read
 
 [dep]
 tiup = to deploy cluster
+expr = to parse `version+path`
```
Comment (innerr) [inline tidb/deploy.bash.ticat:25]: `` => ''

File: helper/helper.bash (hunk 16 above)
Comment (wzru) [reply helper/helper.bash:123]: IMO, here os should be global

File: tidb/deploy.bash [hunk 18]
```diff
@@ -19,4 +19,39 @@ if [ "${skip_exist}" == 'true' ]; then
...
+if [ "${ver_path}" ]; then
+	path="${ver#*+}"
+	ver="${ver_path%+}"
+else
+	path=''
+fi
+
 tiup cluster deploy "${name}" "${ver}" "${yaml}"${confirm}
+
+if [ ${path} ]; then
```
Comment (innerr) [inline tidb/deploy.bash:32]: should be:
```if [ ! -z "${path}" ]```
but this maybe better:
```
if [ -z "${path}" ]; then
    exit
fi
...
```

File: tidb/deploy.bash [hunk 19]
```diff
@@ -19,4 +19,39 @@ if [ "${skip_exist}" == 'true' ]; then
...
+		fi
+		if [ -f "pd-server" ]; then
+			cluster_patch 'pd'
+		fi
+	)
+	elif [ -f "${path}" ]; then
+		base=`basename ${path}`
+		dir=`dirname ${path}`
+		role="${base%*-server}"
+		if [ ${role} ]; then
```
Comment (innerr) [inline tidb/deploy.bash:50]: same as above

File: tidb/deploy.bash [hunk 20]
```diff
@@ -19,4 +19,39 @@ if [ "${skip_exist}" == 'true' ]; then
...
+		(
+			cd ${dir};
```
Comment (innerr) [inline tidb/deploy.bash:52]: quote ""

File: helper/helper.bash [hunk 21]
```diff
@@ -116,3 +116,36 @@ function config_mysql()
...
 }
+
+function must_get_os()
+{
+	if [[ "${OSTYPE}" == 'linux-gnu'* ]]; then
+		local os='linux'
+	elif [[ "${OSTYPE}" == 'darwin'* ]]; then
+		local os='darwin'
+	else
+		echo "[:(] not support os ${OSTYPE}" >&2
```
Comment (innerr) [inline helper/helper.bash:127]: ${OSTYPE} => '${OSTYPE}'

that could make it prettier when OSTYPE is empty

File: tidb/deploy.bash [hunk 22]
```diff
@@ -19,4 +19,39 @@ if [ "${skip_exist}" == 'true' ]; then
...
+	path="${ver#*+}"
+	ver="${ver_path%+}"
+else
+	path=''
+fi
+
 tiup cluster deploy "${name}" "${ver}" "${yaml}"${confirm}
+
+if [ ${path} ]; then
+	if [ -d "${path}" ]; then
```
Comment (innerr) [inline tidb/deploy.bash:33]: this (`if [ -d "${path}"] ... fi`) could also make it a function

**Pull request merged**